File: app/src/main/java/dev/bltucker/nanodegreecapstone/common/sync/DeleteOrphanCommentsAction.java
package dev.bltucker.nanodegreecapstone.common.sync;

import dev.bltucker.nanodegreecapstone.common.data.daos.CommentsDao;
import dev.bltucker.nanodegreecapstone.common.models.Comment;
import io.reactivex.functions.Action;
import timber.log.Timber;

class DeleteOrphanCommentsAction implements Action {

    private final CommentsDao commentsDao;

    DeleteOrphanCommentsAction(CommentsDao commentsDao) {
        this.commentsDao = commentsDao;
    }

    @Override
    public void run() {
        Comment[] rootOrphanComments = commentsDao.getRootOrphanComments();
        int deletedCommentCount = 0;

        for (int i = 0; i < rootOrphanComments.length; i++) {
            Comment parentComment = rootOrphanComments[i];
            Comment[] childComments = commentsDao.getChildComments(parentComment.id);

            deletedCommentCount += commentsDao.deleteComments(childComments);
            deletedCommentCount += commentsDao.deleteComments(parentComment);
        }

        Timber.d("Comment Clean up job removed %d comments", deletedCommentCount);
    }
}

File: app/src/main/java/dev/bltucker/nanodegreecapstone/common/sync/DeleteOrphanCommentsAction.kt
package dev.bltucker.nanodegreecapstone.common.sync

import dev.bltucker.nanodegreecapstone.storydetail.CommentRepository
import io.reactivex.functions.Action
import timber.log.Timber
import javax.inject.Inject

class DeleteOrphanCommentsAction @Inject constructor(private val commentsDao: CommentRepository) : Action {

    override fun run() {
        val rootOrphanComments = commentsDao.getRootOrphanComments()
        var deletedCommentCount = 0

        for (i in rootOrphanComments.indices) {
            val parentComment = rootOrphanComments[i]
            val childComments = commentsDao.getChildComments(parentComment.id)

            deletedCommentCount += commentsDao.deleteComments(*childComments)
            deletedCommentCount += commentsDao.deleteComments(parentComment)
        }

        Timber.d("Comment Clean up job removed %d comments", deletedCommentCount)
    }
}
